imgee/storage.py

```diff
@@ -1,51 +1,77 @@
 # -*- coding: utf-8 -*-
 
-import time
-import os.path
-from subprocess import check_call, CalledProcessError
+from datetime import datetime, timedelta
 from glob import glob
+import os.path
+import os
 import re
-from celery.result import AsyncResult
+from subprocess import check_call, CalledProcessError
+import time
+
 from sqlalchemy import or_
 from werkzeug import secure_filename
 
 import imgee
-from imgee import app, celery
-from imgee.models import db, Thumbnail, StoredFile
-from imgee.async import queueit, get_taskid, BaseTask
-from imgee.utils import (newid, guess_extension, get_file_type,
-                        path_for, get_s3_folder, get_s3_bucket,
-                        download_frm_s3, get_width_height, ALLOWED_MIMETYPES)
+from . import app
+from .models import db, Thumbnail, StoredFile
+from .utils import (
+    newid, guess_extension, get_file_type, is_animated_gif,
+    path_for, get_s3_folder, get_s3_bucket,
+    download_from_s3, get_width_height, ALLOWED_MIMETYPES,
+    exists_in_s3, THUMBNAIL_COMMANDS
+)
 
 
 # -- functions used in views --
 
 def get_resized_image(img, size, is_thumbnail=False):
     """
-    Check if `img` is available with `size` if not make a one. Return the name of it.
+    Check if `img` is available with `size` if not make one. Return the name of it.
     """
+    registry = imgee.registry
     img_name = img.name
+
+    if img.mimetype == 'image/gif':
+        # if the gif file is animated, not resizing it for now
+        # but we will need to resize the gif, keeping animation intact
+        # https://github.com/hasgeek/imgee/issues/55
+        src_path = download_from_s3(img.filename)
+        if is_animated_gif(src_path):
+            return img.name
+
     size_t = parse_size(size)
     if (size_t and size_t[0] != img.width and size_t[1] != img.height) or ('thumb_extn' in ALLOWED_MIMETYPES[img.mimetype] and ALLOWED_MIMETYPES[img.mimetype]['thumb_extn'] != img.extn):
         w_or_h = or_(Thumbnail.width == size_t[0], Thumbnail.height == size_t[1])
         scaled = Thumbnail.query.filter(w_or_h, Thumbnail.stored_file == img).first()
-        if scaled:
+        if scaled and exists_in_s3(scaled):
             img_name = scaled.name
         else:
-            size = get_fitting_size((img.width, img.height), size_t)
+            original_size = (img.width, img.height)
+            size = get_fitting_size(original_size, size_t)
             resized_filename = get_resized_filename(img, size)
-            job = queueit('resize_and_save', img, size, is_thumbnail=is_thumbnail, taskid=resized_filename)
-            return job
+
+            if resized_filename in registry:
+                # file is still being processed
+                # returning None will show "no preview available" thumbnail
```

Comment:
We only have one scenario right now, what to show when an image is processing. so maybe instead of "no preview available" let's just show "loading" spinner. The no preview case might be possible if a file fails to process. but that needs some better handling I think. can I file that as a separate ticket and make this no-preview thumbnail to loading spinner?